codegen/CodeGenerator/src/main/java/AggregateProcessFunctionWriter.java

```diff
@@ -0,0 +1,106 @@
+import org.ainslec.picocog.PicoWriter;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class AggregateProcessFunctionWriter implements ProcessFunctionWriter {
+    private static final String CLASS_NAME = "AggregateProcessFunction";
+    private static final PicoWriter writer = new PicoWriter();
+    private final AggregateProcessFunction aggregateProcessFunction;
+    private final String aggregateType;
+    private final String className;
+
+    public AggregateProcessFunctionWriter(final AggregateProcessFunction aggregateProcessFunction) {
+        requireNonNull(aggregateProcessFunction);
+        this.aggregateProcessFunction = aggregateProcessFunction;
+        Class type = aggregateProcessFunction.getValueType();
+        aggregateType = type.equals(Type.getClass("date")) ? type.getName() : type.getSimpleName();
+        className = aggregateProcessFunction.getName();
+    }
+
+    @Override
+    public void generateCode(String filePath) throws IOException {
+        addImports();
+        addConstructorAndOpenClass();
+        addAggregateFunction();
+        addAdditionFunction();
+        addSubtractionFunction();
+        addInitStateFunction();
+        //class closing
+        writer.writeln_l("}");
+        Files.write(Paths.get(filePath + File.separator + className + ".scala"), writer.toString().getBytes());
+    }
+
+    private void addImports() {
+        writer.writeln("import org.hkust.RelationType.Payload");
+        writer.writeln("import org.apache.flink.api.common.state.ValueStateDescriptor");
+        writer.writeln("import org.apache.flink.api.common.typeinfo.{TypeHint, TypeInformation}");
+        writer.writeln("import org.hkust.BasedProcessFunctions.AggregateProcessFunction");
+    }
+
+    private void addConstructorAndOpenClass() {
+        List<AggregateProcessFunction.AggregateValue> aggregateValues = aggregateProcessFunction.getAggregateValues();
+        writer.writeln_r("class " + className + " extends AggregateProcessFunction[Any, " + aggregateType + "](\"" + className + "\", Array(), Array(),aggregateName = \""
```

Comment:
I think the partition keys should be generated, rather than leaving two "Array()" here.

Reply:
so the 2 `"Array()"` parts are the `this_key` and ``next_key`` in both process and aggregate function?

Reply:
done